core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala

```diff
@@ -133,66 +78,61 @@ class Invoker(
+    override def onMessage(msg: Message)(implicit transid: TransactionId): Future[Any] = {
         require(msg != null, "message undefined")
 
         val start = transid.started(this, LoggingMarkers.INVOKER_ACTIVATION)
         val namespace = msg.action.path
         val name = msg.action.name
-        val version = msg.action.version map { v => DocRevision(v) } getOrElse DocRevision()
-        val action = DocId(WhiskEntity.qualifiedName(namespace, name)).asDocInfo(version)
+        val actionid = DocId(WhiskEntity.qualifiedName(namespace, name)).asDocInfo(msg.revision)
         val tran = Transaction(msg)
         val subject = msg.subject
         val payload = msg.content getOrElse JsObject()
 
-        info(this, s"${action.id} $subject ${msg.activationId}")
+        info(this, s"${actionid.id} $subject ${msg.activationId}")
 
         // caching is enabled since actions have revision id and an updated
         // action will not hit in the cache due to change in the revision id;
         // if the doc revision is missing, then bypass cache
-        WhiskAction.get(entityStore, action.id, action.rev, action.rev != DocRevision()) flatMap {
+        WhiskAction.get(entityStore, actionid.id, actionid.rev, actionid.rev != DocRevision()) recoverWith {
+            case t =>
+                error(this, s"failed to fetch action from db: ${t.getMessage}")
+                Future.failed(ActivationException(s"Failed to fetch action."))
+        } flatMap {
             invokeAction(_, msg.authkey, payload, tran)
+        } recoverWith {
+            // in case of container pull/run operations that fail to execute, assign an appropriate error response
+            case BlackBoxContainerError(msg) => Future.failed(ActivationException(msg, internalError = false))
+            case WhiskContainerError(msg)    => Future.failed(ActivationException(msg))
+            case t: Throwable =>
+                error(this, s"failed during invoke: ${t.getMessage}")
+                Future.failed(ActivationException(s"Failed to run action '${actionid.id}': $msg"))
         } andThen {
-            case Success(activationDoc) =>
-                info(this, s"recorded activation '$activationDoc'")
-            case Failure(t) =>
-                info(this, s"failed to invoke action ${action.id} due to ${t.getMessage}")
-                completeTransactionWithError(action, tran, s"failed to invoke action ${action.id}: ${t.getMessage}")
+            case Failure(ActivationException(e, internal)) =>
+                val response = if (internal) ActivationResponse.whiskError(e) else ActivationResponse.applicationError(e)
+                completeTransactionWithError(actionid.id, msg.action.version, tran, response)
+        } recoverWith {
+            case t: ActivationException =>
+                // for handled exceptions, don't return a failed future
+                Future.successful({})
```

Comment:
adopted - don't need the future.successful since we can flatmap the future from completeTransactionWithError; this is better.